src/awkward/_v2/contents/regulararray.py

```diff
@@ -632,6 +632,85 @@ def _combinations(self, n, replacement, recordlookup, parameters, axis, depth):
                 next, self._size, len(self), self._identifier, self._parameters
             )
 
+    def _reduce_next(
+        self,
+        reducer,
+        negaxis,
+        starts,
+        shifts,
+        parents,
+        outlength,
+        mask,
+        keepdims,
+    ):
+        out = self.toListOffsetArray64(True)._reduce_next(
+            reducer,
+            negaxis,
+            starts,
+            shifts,
+            parents,
+            outlength,
+            mask,
+            keepdims,
+        )
+
+        ### FIXME: need to turn some variable-length dimensions into regular-length dimensions
+
+        # if (!content_.get()->dimension_optiontype()) {
```

Comment:
`dimension_optiontype` is a thing like `purelist_isregular`, `purelist_depth`, `minmax_depth`, etc. They should be implemented directly as properties on Forms:

https://github.com/scikit-hep/awkward-1.0/blob/8aa771b064f6d6b51872390cc2288aed8b952ed8/src/awkward/_v2/forms/listform.py#L144-L173

and then they can be added to all Content classes without code duplication and without creating short-lived Form objects through a Python trick:

https://github.com/scikit-hep/awkward-1.0/blob/8aa771b064f6d6b51872390cc2288aed8b952ed8/src/awkward/_v2/contents/content.py#L746-L764

The "`.__get__(self)`" is to call a Form property using the Content as `self`. This requires duck typing: the implementation of `dimension_optiontype` must work whether `self` is a Form instance or `self` is a Content instance, but for these simple methods, that's usually not hard to do. For instance, the C++ for IndexedForm and IndexedOptionForm are:

https://github.com/scikit-hep/awkward-1.0/blob/8aa771b064f6d6b51872390cc2288aed8b952ed8/src/libawkward/array/IndexedArray.cpp#L142-L145

https://github.com/scikit-hep/awkward-1.0/blob/8aa771b064f6d6b51872390cc2288aed8b952ed8/src/libawkward/array/IndexedArray.cpp#L388-L391

Forms and Contents consistently use `content`/`contents` as a pointer to their child/children, so that should help with the duck typing. For this, there is no reason to ask `isinstance`.

(Now that the mid-layer is being translated into Python, some of the `isinstance` switch statements in the high-level code will be convertible into Content methods, something that I think we'll be slowly doing as post-v2 cleanup.)

Reply:
Done.